README
QRQ - yet another CW trainer - Version 0.3.1

Project website: http://fkurz.net/ham/qrq.html
-----------------------------------------------------------------------------

qrq is an open source Morse telegraphy trainer, similar to the classic DOS
version of Rufz by DL4MM, for Linux, Unix, OS X and Windows. 

It's not intended for learning telegraphy (have a look at radio.linux.org.au
for CW learning software or check out http://lcwo.net/), but to improve the
ability to copy callsigns at high speeds, as needed for example for Contesting. 

-----------------------------------------------------------------------------

COMPILE / INSTALL 

.-----------------------------------------.
| Note: A Windows installer is available  |
|       if you don't like to compile qrq  |
|       by yourself. It can be downloaded |
|       from http://fkurz.net/ham/qrq     |
`-----------------------------------------'

1. make [DESTDIR=/usr] [USE_PA=NO] [USE_WIN32=NO] [USE_CA=NO] [OSX_PLATFORM=NO] [OSX_BUNDLE=YES]

Compiles qrq.

On Linux/Unix, you can compile against Pulse Audio instead of OSS (USE_PA=YES).

If you are building for OSX set OSX_PLATFORM=YES.  If you want to create the
release as an OSX bundle (recommended), then choose OSX_BUNDLE=YES.

To compile it for Windows with MINGW32, using WinMM, use USE_WIN32=YES.

The executable file, 'qrq[.exe]', will be created in the current directory.
To run, qrq only needs the files 'qrqrc', 'toplist', and 'callbase.qcb',
which are also in the current directory. 

At your option, you can install 'qrq' globally:

2. make install [DESTDIR=/usr] [USE_CA=NO] [OSX_PLATFORM=NO] [OSX_BUNDLE=YES]

By default, DESTDIR=/usr, so the executable will be in /usr/bin, the
callsign database in /usr/share/qrq/ and so on.  You can specify any
other destination directory. However, if you build for OSX and have
OSX_BUNDLE=YES, then make install will *not* install into DESTDIR but
will instead make a bundle called qrq.app in the current directory.
You can then drag and drop the qrq.app into your Applications folder if
you wish. 

So to compile for OSX run

make install USE_PA=NO USE_CA=YES OSX_PLATFORM=YES OSX_BUNDLE=YES

and it will create a qrq.app bundle. 

When starting 'qrq' for the first time, it will copy 'qrqrc' and
'toplist' to ~/.qrq/.

3. If you like, edit the configuration file 'qrqrc' according to your needs.
All values can also be changed from within the program.

'qrq' first searches in the current directory. On platforms other than Windows,
~/.qrq/ and then /usr/share/qrq are also searched for these files, if they were
not found in the current directory.

4. If an older version of 'qrq' is already installed, you don't need to remove
it first.

-----------------------------------------------------------------------------
How to use it

Using qrq is simple: qrq sends 50 random calls from a database. After each
call, it waits for the user to enter what he heard and compares the entered
callsign with the one sent. If the callsign is copied correctly, the speed is
increased by 10 CpM and full points are credited, if there were mistakes in the
callsign entered, the speed decreases by 10 CpM and (depending on how many
letters were correct) only a fraction of the maximum points are credited.

A callsign can be heard again once by pressing F6, hitting F10 quits the
attempt. The INS key toggles between insert and overwrite mode in the callsign
field. Pressing F5 leads to the settings screen. The _previous_ callsign
can be reheard by pressing F7.

The possible speed ranges from 20 CpM (4 WpM) to infinity, the initial speed
can be set by the user.

There is a simple toplist function in qrq which makes it possible for the user
to keep track of his training success or to compare scores with others.
You can submit your highscores via e-Mail to fabian@fkurz.net and they will
appear on the toplist published at http://fkurz.net/ham/qrqtop.html.
The toplist is not protected by any kind of checksum, it's based on honesty.

As of version 0.0.7, the toplist file also includes a timestamp of the attempt,
which makes it possible to keep track of your training progress. Pressing F7
generates a graph score vs. date (GNUplot required, not on Windows).

Options can be changed in the config file qrqrc or via the options menu (F5).
As of version 0.2.0, some additional training modes are available, which allow
e.g. unlimited usage of F6 (call repetition) and attempts that are longer than
the normal 50 calls.

A small Perl script, qrqscore, to synchronize the online-toplist
(http://fkurz.net/ham/qrqtop.php) with your local toplist is included as of
version 0.1.2.

-----------------------------------------------------------------------------

Download, License

Of course qrq is free software (free as in beer and free as in freedom) and
published under the GPL 2.

If you wish to use the files coreaudio.h and coreaudio.c in a project separate
from qrq, they are licensed under the MIT license.

-----------------------------------------------------------------------------

Contact, Feedback

I am always interested in any kind of feedback concerning qrq.
If you have any suggestions, questions, feature-requests etc., don't hesitate
a minute and contact the author: Fabian Kurz, DJ1YFK <fabian@fkurz.net>.


ChangeLog
0.3.1 / 2013-01-06   Fabian Kurz <fabian@fkurz.net>
	* Fixed saving scores if toplist file is not in cwd (tnx ON4ACP)
	* F7 repeats the _previous_ callsign/word (suggested by KU1T)
	* The sample rate can be set via qrqrc now.
	* Minor bugfixes in PulseAudio output

0.3.0 / 2011-12-18   Fabian Kurz <mail@fkurz.net>
	* New platform: Windows (with MINGW), thanks to SP8QED for patches
	* CW Settings can be changed during attempts on the fly (F5)
	* Attempts can be aborted without aborting the whole program
	* A pretty new icon (thanks to Victoria, SV2KBS)

0.2.1 / 2011-09-19   Fabian Kurz <mail@fkurz.net>
	* Added support for native CoreAudio on OSX.  Also builds for iOS.
	  [tnx KB1OOO]
	* Removed support for OpenAL now that native OSX is supported and Pulse
	  Audio is available. [tnx KB1OOO]
	* Added English word list english.qcb  [tnx Chuck Vaughn, AA0HW]
	* Fixed missing audio on second channel for OSS
	* Fixed Debian Bug 771075 [tnx John Nogatch, AC6SL]

0.2.0 / 2011-05-07   Fabian Kurz <mail@fkurz.net>
	* Added options for fixed CW speed (regardless of errors), unlimited
	usage of repeats and unlimited attempt lengths. Suggested by AA0HW.
	* No more duplicate callsigns in one attempt
	* Unlimited attempts (e.g. with _all_ callsigns of a particular
	  database) added
	* Call database file can be changed/selected in a menu now
	* Added a database with the 100 most common English words
	* Call databases with DOS linebreaks can now be used.
	* Fixed bugs in memory allocation.
	* CW edges changed to raised cosine
	* Minimum character speed can be specified; below, Farnsworth is used
	* PulseAudio support added (OSS and OpenAL still available)
	* Reformatted ChangeLog

0.1.4 / 2008-06-22   Fabian Kurz <mail@fkurz.net>
	* Implemented dynamic allocation of the callsign database array.
	* Callsign database updated, over 22000 callsigns now.
	* Added support for building under OpenAL.
	* Native build option for OSX; tnx KB1OOO!

0.1.3 / 2007-12-15   Fabian Kurz <mail@fkurz.net>  
	* Fixed handling of F- and Cursor-Keys on UTF-8 systems; 
	Home-/End-Keys now also implemented.
	* Some minor optical changes.

0.1.2 / 2007-11-17   Fabian Kurz <mail@fkurz.net>
	* Speeds are displayed in both WpM and CpM now
	* Different CW waveforms (sine, sawtooth, squarewave) are now available 
	* Backspace should now work in all terminals
	* Added a Perl script ('qrqscore') to synchronize your toplist with
	the internet toplist
	* Added an ugly desktop icon

0.1.1 / 2007-07-08   Fabian Kurz <mail@fkurz.net>
	* qrq now ensures that it can write to 'qrqrc' and 'toplist' after
	copying it from DESTDIR/share/qrq/ to ~/.qrq/
	* Added all calls of the author's current log to the callsign database;
	now 20084 unique callsigns.

0.1.0 / 2007-06-30   Fabian Kurz <mail@fkurz.net>
	* User can chose between random CW tone pitch (like before) and a fixed
	frequency now, which can be set in qrqrc or in the config menu. Tnx
	DL9PF fr patch!

0.0.7 / 2007-06-17   Fabian Kurz <mail@fkurz.net>
	* Added a timestamp to the toplist-file, which makes it possible to
	track your training progress.
	* Pressing F7 will generate a graph of score vs.  date with gnuplot
	(if installed).

0.0.6 / 2007-06-16   Fabian Kurz <mail@fkurz.net>
	* Added Makefile. Program looks for its files in different directories
	now and creates ~/.qrq/ unless toplist and qrqrc are in the current
	directory.

0.0.5 / 2006-07-29   Fabian Kurz <mail@fkurz.net>
	* Settings (speed, mycall, risetimes..) can be changed from within the
	program now (F5).

0.0.4 / 2006-07-14   Fabian Kurz <mail@fkurz.net>
	* Bugfix. Waiting for a non-existing thread caused segfaults. 
	Thanks DL5RB!

0.0.3 / 2006-06-25   Fabian Kurz <mail@fkurz.net>
	* The CW output is now in a separate thread, making it possible to see
	what the user typed during CW output.
	* Added a sanity check of values read from the configuration file. It
	shouldn't be possible to cause any problems by invalid entries anymore.
	* Several minor bugfixes.

0.0.2 / 2006-06-23   Fabian Kurz <mail@fkurz.net>
	* Changed scoring system, other minor changes.

0.0.1 / 2006-05-19   Fabian Kurz <mail@fkurz.net>
	* Initial release


coreaudio.c
/*
Copyright (c) 2011 Marc Vaillant

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#include <AudioToolbox/AudioToolbox.h>
#include <AudioUnit/AudioUnit.h>
#include <pthread.h>


#define kOutputBus 0
#define kInputBus 1
#define SAMPLE_RATE 44100

static int* _pcm;
static int _pcmSize;
static int _index;
static AudioComponentInstance* _audioUnit = 0;
static pthread_mutex_t _playingMutex;
static pthread_cond_t _playingCond;


void write_audio(void* dummy, int* pcm, int size)
{

	_pcm = pcm;
	_pcmSize = size / sizeof(int);
	_index = 0;
	pthread_mutex_lock(&_playingMutex);
  AudioOutputUnitStart(*_audioUnit);

	// block until finished playing
	pthread_cond_wait(&_playingCond, &_playingMutex);
	pthread_mutex_unlock(&_playingMutex);

}

static OSStatus playbackCallback(void *inRefCon, 
                                  AudioUnitRenderActionFlags *ioActionFlags, 
                                  const AudioTimeStamp *inTimeStamp, 
                                  UInt32 inBusNumber, 
                                  UInt32 inNumberFrames, 
                                  AudioBufferList *ioData) 
{    
	//    cout<<"index = "<<_index<<endl;
	//	cout<<"numBuffers = "<<ioData->mNumberBuffers<<endl;

	//int totalNumberOfSamples = _pcm.size();
	int totalNumberOfSamples = _pcmSize;
	for(UInt32 i = 0; i < ioData->mNumberBuffers; ++i)
	{
		//      cout<<"i = "<<i<<endl;
		int samplesLeft = totalNumberOfSamples - _index;
		int numSamples = ioData->mBuffers[i].mDataByteSize / 4;
		//	  cout<<"numSamples = "<<numSamples<<endl;
		if(samplesLeft > 0)
		{
			if(samplesLeft < numSamples)
			{
				memcpy(ioData->mBuffers[i].mData, &_pcm[_index], samplesLeft * 4);
				_index += samplesLeft;
				memset(((char*) (ioData->mBuffers[i].mData)) + samplesLeft * 4, 0, (numSamples - samplesLeft) * 4) ;
			}
			else
			{
				memcpy(ioData->mBuffers[i].mData, &_pcm[_index], numSamples * 4) ;
				_index += numSamples;
			}

			//		ofstream fp("buffer.txt", ios::app);
			//		for(UInt32 i = 0; i < ioData->mBuffers[0].mDataByteSize / 4; ++i)
			//		{
			//		  fp<<*((short*) ioData->mBuffers[0].mData + 2 * i)<<endl;
			//		}
			//		fp.close();


		}
		else
		{
			memset(ioData->mBuffers[i].mData, 0, ioData->mBuffers[i].mDataByteSize);

			// signal that pcm is finished playing
			pthread_mutex_lock(&_playingMutex);
			pthread_cond_signal(&_playingCond);
			pthread_mutex_unlock(&_playingMutex);

			// stop the audio unit
			AudioOutputUnitStop(*_audioUnit);
		}
	}


	return noErr;
}

void close_audio(void* cookie)
{
	AudioOutputUnitStop(*_audioUnit);
}

void close_dsp(void* s)
{
	AudioUnitUninitialize(*_audioUnit);
	AudioComponentInstanceDispose(*_audioUnit);
	free(_audioUnit);
	_audioUnit = 0;
	pthread_mutex_destroy(&_playingMutex);
  pthread_cond_destroy (&_playingCond);
}

void* open_dsp(char* dummy)
{

	// skip if already created audio unit
	
	if(_audioUnit)
		return 0;

	pthread_mutex_init(&_playingMutex, NULL);
  pthread_cond_init (&_playingCond, NULL);
	
	_audioUnit = (AudioComponentInstance*) malloc(sizeof(AudioComponentInstance));
  OSStatus status;
  //AudioComponentInstance audioUnit;

  // Describe audio component
  AudioComponentDescription desc;
  desc.componentType = kAudioUnitType_Output;
#ifdef __IPHONE_OS_VERSION_MIN_REQUIRED
  desc.componentSubType = kAudioUnitSubType_RemoteIO;
#else
  desc.componentSubType = kAudioUnitSubType_DefaultOutput;
#endif
  desc.componentFlags = 0;
  desc.componentFlagsMask = 0;
  desc.componentManufacturer = kAudioUnitManufacturer_Apple;

  // Get component
  AudioComponent inputComponent = AudioComponentFindNext(NULL, &desc);

  // Get audio units
  status = AudioComponentInstanceNew(inputComponent, _audioUnit);
  //checkStatus(status);

  UInt32 flag = 1;
  // Enable IO for playback
  status = AudioUnitSetProperty(*_audioUnit, 
				  kAudioOutputUnitProperty_EnableIO, 
				  kAudioUnitScope_Output, 
				  kOutputBus,
				  &flag, 
				  sizeof(flag));
  //checkStatus(status);

  // Describe format

  AudioStreamBasicDescription audioFormat;
  audioFormat.mSampleRate = SAMPLE_RATE;
  audioFormat.mFormatID	= kAudioFormatLinearPCM;
  audioFormat.mFormatFlags = kAudioFormatFlagIsSignedInteger | kAudioFormatFlagIsPacked;
  audioFormat.mFramesPerPacket = 1;
  audioFormat.mChannelsPerFrame = 2;
  audioFormat.mBitsPerChannel = 16;
  audioFormat.mBytesPerPacket = 4;
  audioFormat.mBytesPerFrame = 4;

  // Apply format

  status = AudioUnitSetProperty(*_audioUnit, 
				  kAudioUnitProperty_StreamFormat, 
				  kAudioUnitScope_Input, 
				  kOutputBus, 
				  &audioFormat, 
				  sizeof(audioFormat));
//  checkStatus(status);

  // Set output callback
  AURenderCallbackStruct callbackStruct;
  callbackStruct.inputProc = playbackCallback;
  callbackStruct.inputProcRefCon = NULL;
  status = AudioUnitSetProperty(*_audioUnit, 
				  kAudioUnitProperty_SetRenderCallback, 
				  kAudioUnitScope_Global, 
				  kOutputBus,
				  &callbackStruct, 
				  sizeof(callbackStruct));

  // Initialize
  status = AudioUnitInitialize(*_audioUnit);

	return 0;

}

//int main()
//{    
//
//  //generate pcm tone  freq = 800, duration = 5s, rise/fall time = 5ms
//
//  generateTone(_pcm, 800, 500, SAMPLE_RATE, 5, 0.5);
//  _index = 0;
//
//  OSStatus status;
//  AudioComponentInstance audioUnit;
//
//  // Describe audio component
//  AudioComponentDescription desc;
//  desc.componentType = kAudioUnitType_Output;
//  desc.componentSubType = kAudioUnitSubType_DefaultOutput;
//  desc.componentFlags = 0;
//  desc.componentFlagsMask = 0;
//  desc.componentManufacturer = kAudioUnitManufacturer_Apple;
//
//  // Get component
//  AudioComponent inputComponent = AudioComponentFindNext(NULL, &desc);
//
//  // Get audio units
//  status = AudioComponentInstanceNew(inputComponent, &audioUnit);
//  //checkStatus(status);
//
//  UInt32 flag = 1;
//  // Enable IO for playback
//  status = AudioUnitSetProperty(audioUnit, 
//				  kAudioOutputUnitProperty_EnableIO, 
//				  kAudioUnitScope_Output, 
//				  kOutputBus,
//				  &flag, 
//				  sizeof(flag));
//  //checkStatus(status);
//
//  // Describe format
//
//  AudioStreamBasicDescription audioFormat;
//  audioFormat.mSampleRate = SAMPLE_RATE;
//  audioFormat.mFormatID	= kAudioFormatLinearPCM;
//  audioFormat.mFormatFlags = kAudioFormatFlagIsSignedInteger | kAudioFormatFlagIsPacked;
//  audioFormat.mFramesPerPacket = 1;
//  audioFormat.mChannelsPerFrame = 2;
//  audioFormat.mBitsPerChannel = 16;
//  audioFormat.mBytesPerPacket = 4;
//  audioFormat.mBytesPerFrame = 4;
//
//  // Apply format
//
//  status = AudioUnitSetProperty(audioUnit, 
//				  kAudioUnitProperty_StreamFormat, 
//				  kAudioUnitScope_Input, 
//				  kOutputBus, 
//				  &audioFormat, 
//				  sizeof(audioFormat));
////  checkStatus(status);
//
//  // Set output callback
//  AURenderCallbackStruct callbackStruct;
//  callbackStruct.inputProc = playbackCallback;
//  callbackStruct.inputProcRefCon = NULL;
//  status = AudioUnitSetProperty(audioUnit, 
//				  kAudioUnitProperty_SetRenderCallback, 
//				  kAudioUnitScope_Global, 
//				  kOutputBus,
//				  &callbackStruct, 
//				  sizeof(callbackStruct));
//
//  // Initialize
//  status = AudioUnitInitialize(audioUnit);
//  status = AudioOutputUnitStart(audioUnit);
//
//  // sleep
//  
//  sleep(2.0);
//  return 1;
//}



coreaudio.h
/*
Copyright (c) 2011 Marc Vaillant

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

*/

#ifndef CORE_AUDIO_IMP
#define CORE_AUDIO_IMP

#ifdef __cplusplus
extern "C" 
{
#endif

  void* open_dsp(char * dummy);
  int write_audio(void* cookie, int* incomingBuffer, int incomingSize);
  int close_audio(void* cookie);

#ifdef __cplusplus
}
#endif

#endif



pulseaudio.h
/* 
Copyright (C) 2011  Fabian Kurz

$Id: pulseaudio.h 203 2011-03-21 22:55:21Z dj1yfk $

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
Street, Fifth Floor, Boston, MA  02110-1301, USA.

*/ 

#ifndef QRQ_PA
#define QRQ_PA

void *open_dsp (); 
void write_audio (void *bla, int *in, int size);
void close_audio (void *s);

#endif


pulseaudio.c
/* 
Copyright (C) 2011  Fabian Kurz, DJ1YFK
 
$Id: pulseaudio.c 541 2012-08-17 17:58:26Z dj1yfk $

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
Street, Fifth Floor, Boston, MA  02110-1301, USA.

PulseAudio specific functions and includes.

*/

#include <ncurses.h>
#include <stdlib.h>
#include <fcntl.h>
#include <pulse/simple.h>
#include <pulse/error.h>

extern long samplerate;
extern void  *dsp_fd;

short int buf[441000];		/* 10 secs buffer */
int bufpos = 0;

void *open_dsp () {
	static int opened = 0;

	/* with PA we only open the device once and then leave it
	   opened */
	if (opened) {
		return dsp_fd;
	}

	/* The Sample format to use */
	static pa_sample_spec ss = {
		.format = PA_SAMPLE_S16LE,
		.rate = 8000,
		.channels = 1
	};
	ss.rate = samplerate;
	pa_simple *s = NULL;
	int error;

	if (!(s = pa_simple_new(NULL, "qrq", PA_STREAM_PLAYBACK, NULL, 
				"playback", &ss, NULL, NULL, &error))) {
	        fprintf(stderr, "pa_simple_new() failed: %s\n", 
				pa_strerror(error));
	}

	opened = 1;
	return s;
}

/* actually just puts samples into the buffer that is played at the end 
(close_audio) */
void write_audio (void *bla, int *in, int size) {
	int i = 0;
	for (i=0; i < size/sizeof(int); i++) {
		buf[bufpos] = (short int) in[i];
		bufpos++;
	}	
}

void close_audio (void *s) {
	int e;
	pa_simple_write(s, buf, bufpos*sizeof(short int), &e);
	pa_simple_drain(s, &e);
	bufpos = 0;
}


